ta/pkcs11/src/processing_asymm.c

```diff
@@ -0,0 +1,556 @@
+				   struct pkcs11_object *obj)
+{
+	static const uint32_t pkcs2tee_algo[][3] = {
+		/* EC flavors (Must find key size from the object) */
+		{ PKCS11_CKM_ECDSA, 1, 0 },
+		{ PKCS11_CKM_ECDSA_SHA1, 1, TEE_ALG_SHA1 },
+		{ PKCS11_CKM_ECDSA_SHA224, 1, TEE_ALG_SHA224 },
+		{ PKCS11_CKM_ECDSA_SHA256, 1, TEE_ALG_SHA256 },
+		{ PKCS11_CKM_ECDSA_SHA384, 1, TEE_ALG_SHA384 },
+		{ PKCS11_CKM_ECDSA_SHA512, 1, TEE_ALG_SHA512 },
+	};
+	size_t end = sizeof(pkcs2tee_algo) / (2 * sizeof(uint32_t));
+	size_t n = 0;
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+
+	for (n = 0; n < end; n++) {
+		if (proc_params->id == pkcs2tee_algo[n][0]) {
+			*tee_id = pkcs2tee_algo[n][1];
+			*tee_hash_id = pkcs2tee_algo[n][2];
+			break;
+		}
+	}
+
+	if (n == end)
+		return PKCS11_RV_NOT_IMPLEMENTED;
+
+	switch (proc_params->id) {
+	case PKCS11_CKM_ECDSA:
+	case PKCS11_CKM_ECDSA_SHA1:
+	case PKCS11_CKM_ECDSA_SHA224:
+	case PKCS11_CKM_ECDSA_SHA256:
+	case PKCS11_CKM_ECDSA_SHA384:
+	case PKCS11_CKM_ECDSA_SHA512:
+		rc = pkcs2tee_algo_ecdsa(tee_id, proc_params, obj);
+		break;
+	default:
+		rc = PKCS11_CKR_OK;
+		break;
+	}
+
+	return rc;
+}
+
+static uint32_t pkcs2tee_key_type(uint32_t *tee_type, struct pkcs11_object *obj,
+				  enum processing_func function)
+{
+	enum pkcs11_class_id class = get_class(obj->attributes);
+	enum pkcs11_key_type type = get_key_type(obj->attributes);
+
+	switch (class) {
+	case PKCS11_CKO_PUBLIC_KEY:
+	case PKCS11_CKO_PRIVATE_KEY:
+		break;
+	default:
+		TEE_Panic(class);
+		break;
+	}
+
+	switch (type) {
+	case PKCS11_CKK_EC:
+		assert(function != PKCS11_FUNCTION_DERIVE);
+
+		if (class == PKCS11_CKO_PRIVATE_KEY)
+			*tee_type = TEE_TYPE_ECDSA_KEYPAIR;
+		else
+			*tee_type = TEE_TYPE_ECDSA_PUBLIC_KEY;
+		break;
+	default:
+		TEE_Panic(type);
+		break;
+	}
+
+	return PKCS11_CKR_OK;
+}
+
+static uint32_t allocate_tee_operation(struct pkcs11_session *session,
+				       enum processing_func function,
+				       struct pkcs11_attribute_head *params,
+				       struct pkcs11_object *obj)
+{
+	uint32_t size = (uint32_t)get_object_key_bit_size(obj);
+	uint32_t algo = 0;
+	uint32_t hash_algo = 0;
+	uint32_t mode = 0;
+	uint32_t hash_mode = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct active_processing *processing = session->processing;
+
+	assert(processing->tee_op_handle == TEE_HANDLE_NULL);
+	assert(processing->tee_hash_op_handle == TEE_HANDLE_NULL);
+
+	if (pkcs2tee_algorithm(&algo, &hash_algo, function, params, obj))
+		return PKCS11_CKR_FUNCTION_FAILED;
+
+	pkcs2tee_mode(&mode, function);
+
+	if (hash_algo) {
+		pkcs2tee_mode(&hash_mode, PKCS11_FUNCTION_DIGEST);
+
+		res = TEE_AllocateOperation(&processing->tee_hash_op_handle,
+					    hash_algo, hash_mode, 0);
+		if (res) {
+			EMSG("TEE_AllocateOp. failed %#"PRIx32" %#"PRIx32,
+			     hash_algo, hash_mode);
+
+			TEE_FreeOperation(processing->tee_hash_op_handle);
+			processing->tee_hash_op_handle = TEE_HANDLE_NULL;
+			return PKCS11_CKR_MECHANISM_INVALID;
```

Comment:
I think return `tee2pkcs_error(res);` is more appropriate.
Or maybe:
```c
	if (res == TEE_ERROR_NOT_SUPPORTED)
		return PKCS11_CKR_MECHANISM_INVALID;
	return tee2pkcs_error(res);

Reply:
Snippet looks OK and matches the functionality for `tee_op_handle`.

Reply:
This is called on failure of `TEE_AllocateOperation()`. It may be some issue not related to the mechanism being invalid. So we should return the real reason why the operation failed.

We really want to return `PKCS11_CKR_MECHANISM_INVALID` is the mechanism is not supported, hence I suggested a specific treatment for `TEE_ERROR_NOT_SUPPORTED`. As for the other error codes, I think `tee2pkcs_error(res)` would do the job.